File: Difficulty 4.1/Rock Paper Scissors/RockPaperScissors.java
import java.util.Scanner;
import java.util.HashMap;
import java.util.Iterator;
import java.lang.Double;

public class RockPaperScissors{
    public static void main (String [] args) {
        Scanner sc = new Scanner(System.in);
        int counter = 0;
        while (true) {
            int players = sc.nextInt();
            if (players == 0) {
                break;
            }

            if (counter != 0) {
                System.out.println();
            }
            
            HashMap<Integer, Player> list = new HashMap<>();
            int games = sc.nextInt();

            for (int i = 0; i < players; i++) {
                list.put(i + 1, new Player());
            }
            
            for (int i = 0; i < games; i++) {
                int one = sc.nextInt();
                String onePlay = sc.next();
                int two = sc.nextInt();
                String twoPlay = sc.next();

                if (onePlay.equals("rock") && twoPlay.equals("scissors")) {
                    Player p1 = list.get(one);
                    Player p2 = list.get(two);
                    p1.addWin();
                    p2.addLose();
                    list.replace(one, p1);
                    list.replace(two, p2);
                } else if (onePlay.equals("scissors") && twoPlay.equals("paper")) {
                    Player p1 = list.get(one);
                    Player p2 = list.get(two);
                    p1.addWin();
                    p2.addLose();
                    list.replace(one, p1);
                    list.replace(two, p2);
                } else if (onePlay.equals("paper") && twoPlay.equals("rock")) {
                    Player p1 = list.get(one);
                    Player p2 = list.get(two);
                    p1.addWin();
                    p2.addLose();
                    list.replace(one, p1);
                    list.replace(two, p2);
                } else if (onePlay.equals("paper") && twoPlay.equals("scissors")) {
                    Player p1 = list.get(one);
                    Player p2 = list.get(two);
                    p1.addLose();
                    p2.addWin();
                    list.replace(one, p1);
                    list.replace(two, p2);
                } else if (onePlay.equals("scissors") && twoPlay.equals("rock")) {
                    Player p1 = list.get(one);
                    Player p2 = list.get(two);
                    p1.addLose();
                    p2.addWin();
                    list.replace(one, p1);
                    list.replace(two, p2);
                } else if (onePlay.equals("rock") && twoPlay.equals("paper")) {
                    Player p1 = list.get(one);
                    Player p2 = list.get(two);
                    p1.addLose();
                    p2.addWin();
                    list.replace(one, p1);
                    list.replace(two, p2);
                } else {

                }
            }

            Iterator it = list.values().iterator();

            while(it.hasNext()) {
                Player currPlayer = (Player) it.next();
                try {
                    if (currPlayer.wins + currPlayer.loses == 0) {
                        System.out.println("-");
                    } else {
                        double result = (double) currPlayer.wins / (double) (currPlayer.wins + currPlayer.loses);
                        System.out.printf("%.3f", result);
                        System.out.println();
                    }
                } catch (Exception e) {
                    System.out.println("-");
                }
            }
            counter++;
        }
    }
}

class Player {
    int wins = 0;
    int loses = 0;

    public void addWin() {
        wins++;
    }

    public void addLose() {
        loses++;
    }
}

File: Difficulty 4.1/Rock Paper Scissors/RockPaperScissors.py
counter = 0

def isDraw(onePlay, twoPlay):
    if onePlay == twoPlay:
        return True
    return False

def isWin(onePlay, twoPlay):
    if onePlay == "rock" and twoPlay == "scissors":
        return True
    elif onePlay == "scissors" and twoPlay == "paper":
        return True
    elif onePlay == "paper" and twoPlay == "rock":
        return True
    else:
        return False

while True:
    line = input().split(" ")
    players = int(line[0])
    if players == 0:
        break

    if counter != 0:
        print()

    wins = []
    loses = []

    for x in range(players + 1):
        wins.append(0)
        loses.append(0)

    games = int(int(line[1]) * players * (players - 1) / 2)

    for x in range(games):
        line = input().split(" ")
        one = int(line[0])
        onePlay = line[1]
        two = int(line[2])
        twoPlay = line[3]

        if isDraw(onePlay, twoPlay):
            pass
        elif isWin(onePlay, twoPlay):
            wins[one] += 1
            loses[two] += 1
        else:
            loses[one] += 1
            wins[two] += 1

    for x in range(1, len(wins)):
        winner = wins[x]
        loser = loses[x]

        if winner + loser == 0:
            print("-")
        else:
            print("{0:.3f}".format((winner / (winner + loser))))
    
    counter += 1